DiffMatchPatch.zig

```diff
@@ -96,10 +129,15 @@ fn diffInternal(
     deadline: u64,
 ) DiffError!DiffList {
     // Check for equality (speedup).
-    var diffs = DiffList{};
     if (std.mem.eql(u8, before, after)) {
+        var diffs = DiffList{};
+        errdefer deinitDiffList(allocator, &diffs);
         if (before.len != 0) {
-            try diffs.append(allocator, Diff.init(.equal, try allocator.dupe(u8, before)));
+            try diffs.ensureUnusedCapacity(allocator, 1);
+            diffs.appendAssumeCapacity(Diff.init(
+                .equal,
+                try allocator.dupe(u8, before),
+            ));
```

Comment:
Just a minor style suggestion, but when appending 1 item, I prefer to use `append()`.  `ensureTotalCapacity() + appendAssumeCapacity()` is just more verbose and does the same thing as `append()`.

Please ignore me if this was already mentioned or counter to @Techatrix reviews.  I have been following this thread but may have missed some of it.

Reply:
😅 There was a whole discussion, it's gone because of the changes.

The problem, which @Techatrix identified, is that just using `append()` means that the duped text leaks if the append fails. He graciously reworked the tests to catch all `OutOfMemory` errors, and status quo is that I've fixed the leaks and double-frees that revealed. 

It was a real learning experience for me, an eye-opener. I'm experienced with C, but Zig' execution model (because of defers, and how errors work) is actually quite different! Making sure that every sequence between any `try` expression has a correct memory policy is a whole different lens through which to view the code.

Reply:
My suggestion was to use `append()` instead _only_ when there is a single `append()`.  I did see Techatrix's advice on using `ensureTotalCapacity() + appendAssumeCapacity()`.  But afaik that only made sense when there were 2 or more appends.

Reply:
Nope, you still need to separate them. It's because `appendAssumeCapacity` doesn't use a `try`, so it can't fail. The pattern isn't an efficiency thing (using it for several inserts/appends is more efficient though), it's a memory safety thing.

What this avoids: the `try allocator.dupe` succeeds, but the `try append`  fails. That means the new `[]const u8` doesn't have anywhere to live, so it leaks. 

I considered rewriting `diffAppend` and `diffInsert` to be inline and do the `ensureUnused`/`AssumeCapacity` dance where we don't have to look at it. But I figured that could be interpreted as stylistic bickering, and I'm more interested in landing the patch than on resolving a difference like that.

And I'm glad I didn't. There are a number of hairy spots in the code where those two operations have to be ordered in a way where something else comes between them.

Reply:
As an aside, I feel that the standard library could be improved by replacing every use of `Capacity` with `Space`. I don't see how that could be taken ambiguously, and it makes the calls three letters, and two syllables, shorter. 

The difference isn't huge, but every column of width comes at a premium cost.  In the meantime, I'm willing to pay the price of verbosity for the reward of robust code.

Reply:
> What this avoids: the `try allocator.dupe` succeeds, but the `try append` fails. That means the new `[]const u8` doesn't have anywhere to live, so it leaks.

Ah!  Got it now.  Thanks for clearing this up for me.  I wasn't considering an orphaned dupe().